pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -284,6 +286,58 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
 
 	ipAddrNetworkNames := toIPAddrNetworkNames(nonVNICDevices)
 	nonLocalhostIPs := excludeLocalhostIPs(ipAddrNetworkNames)
+	guestInfo, encoding := extractGuestInfo(oVM.Config.ExtraConfig)
+
+	if guestInfo != "" && encoding == "base64" {
+		value, err := base64.StdEncoding.DecodeString(guestInfo)
+		if err != nil {
+			klog.Errorf("Error decoding guestinfo for vm=%+v in vc=%s and datacenter=%s: %v",
+				vmDI.VM, vmDI.VcServer, vmDI.DataCenter.Name(), err)
+			return err
+		}
+
+		type cloudInitConfig struct {
+			Network struct {
+				Ethernets map[string]struct {
+					Name      string   `yaml:"set-name"`
+					Addresses []string `yaml:"addresses"`
+				} `yaml:"ethernets"`
+			} `yaml:"network"`
+		}
+
+		guestInfo := &cloudInitConfig{}
+		err = yaml.Unmarshal(value, guestInfo)
+		if err != nil {
+			klog.Errorf("Error unmarshaling guestinfo for vm=%+v in vc=%s and datacenter=%s: %v",
+				vmDI.VM, vmDI.VcServer, vmDI.DataCenter.Name(), err)
+			return err
+		}
+
+		guestInfoAddresses := []net.IP{}
+		for _, eth := range guestInfo.Network.Ethernets {
+			for _, address := range eth.Addresses {
+				guestInfoAddresses = append(
+					guestInfoAddresses,
+					net.ParseIP(strings.Split(address, "/")[0]))
+			}
+		}
+
+		// We want to prefer addresses that are from the guestInfo but only if
+		// they are on a nic already
+		ips := []*ipAddrNetworkName{}
```

Comment:
#L327 to #L337 is trying to find the `guestInfoAddress` in `nonLocalhostIPs`:

 I'd suggest that we use a `map[net.IP]bool` to store the `guestInfoAddresses` and check if `nonLocalhostIPs` contains the IP or not which can reduce the time complexity and make the code easier to understand.

Reply:
We decided to rewrite the code to prioritize readability so we changed the implementation to a sort to better describe the intent and extracted the code into a separate function.